### DISCUSSION TITLE
nextjs middleware issue while calling sequelize model into it

### DISCUSSION THREAD

**Discussion by Sanskar696** [vercel/next.js#50587] [Help] [answered]
### Summary

![image](https://github.com/vercel/next.js/assets/125652661/fef5b81e-4613-4389-892e-b356c771ca57)

Hey there Can any one help me out 

- warn ./node_modules/sequelize/lib/dialects/abstract/connection-manager.js
Critical dependency: the request of a dependency is an expression

Import trace for requested module:
./node_modules/sequelize/lib/dialects/abstract/connection-manager.js
./node_modules/sequelize/lib/dialects/mariadb/connection-manager.js
./node_modules/sequelize/lib/dialects/mariadb/index.js
./node_modules/sequelize/lib/sequelize.js
./node_modules/sequelize/lib/index.js
./node_modules/sequelize/lib/index.mjs
./models/init_models.js

./node_modules/sequelize/lib/dialects/abstract/connection-manager.js
Critical dependency: the request of a dependency is an expression

Import trace for requested module:
./node_modules/sequelize/lib/dialects/abstract/connection-manager.js
./node_modules/sequelize/lib/dialects/mariadb/connection-manager.js
./node_modules/sequelize/lib/dialects/mariadb/index.js
./node_modules/sequelize/lib/sequelize.js
./node_modules/sequelize/lib/index.js
./node_modules/sequelize/lib/index.mjs
./models/init_models.js
- error node_modules/sequelize/lib/utils/deprecations.js (34:0) @ <unknown>
- error (0 , import_util.deprecate) is not a function

**Comment by mubarakn** [comment]
I'm also experiencing exactly the same above error

**Comment by mubarakn** [comment]
have you found any solution @Sanskar696 ?

**Reply by Sanskar696** [reply]
Not yet @mubarakn

**Comment by ferbragon** [comment]
I'm have exactly the same above error:

./node_modules/sequelize/lib/dialects/abstract/connection-manager.js
Critical dependency: the request of a dependency is an expression

Import trace for requested module:
./node_modules/sequelize/lib/dialects/abstract/connection-manager.js
./node_modules/sequelize/lib/dialects/mariadb/connection-manager.js
./node_modules/sequelize/lib/dialects/mariadb/index.js
./node_modules/sequelize/lib/sequelize.js
./node_modules/sequelize/lib/index.js
./node_modules/sequelize/lib/index.mjs
./src/app/api/hello/route.ts

**Comment by Lalman888** [comment]
I also have same error

**Comment by gerardnico** [comment]
If you google it:
[Critical dependency: the request of a dependency is an expression](https://www.google.com/search?q=Critical+dependency%3A+the+request+of+a+dependency+is+an+expression&sourceid=chrome&ie=UTF-8)
, you see that it is a webpack error due to a dynamic import (ie `require('' + 'nodent')` in:
```
./node_modules/sequelize/lib/dialects/abstract/connection-manager.js
```

You can try to:
* upgrade it
* exclude it from the webpack nextjs configuration.

Here, a thread on [webpack and sequelize](https://github.com/sequelize/sequelize/issues/13169)

**Comment by ferbragon** [comment]
I think that is a Next config issue. If anyone else has this problem, the solution is to add serverComponentsExternalPackages: ["sequelize"] to the next.config.js like this: 

/** @type {import('next').NextConfig} */
const nextConfig = {
    experimental: {
      serverComponentsExternalPackages: ['sequelize'],
    },
  }
   
  module.exports = nextConfig

as documented here: https://nextjs.org/docs/app/api-reference/next-config-js/serverComponentsExternalPackages.

I would appreciate it if you could provide an alternative solution or if you think this may not be the ideal solution for your project

**Reply by ccarapezza** [reply]
In my case, i am using sequelize-typescript and I also had to add it to the list.

```
const nextConfig = {
    experimental: {
        serverComponentsExternalPackages: ['sequelize', 'sequelize-typescript'],
    }
}

module.exports = nextConfig
```

**Reply by Jan-T-Berg** [reply]
TY! This fixxed my problem. Thanks.

**Reply by ayushmina** [reply]
have you solved this error?

**Reply by clockback** [reply]
Update, now should be:
```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  serverExternalPackages: ['sequelize']
}

module.exports = nextConfig
```

**Reply by vinaydanidhariya** [reply]
> ```js
> /** @type {import('next').NextConfig} */
> const nextConfig = {
>   serverExternalPackages: ['sequelize']
> }
> ```

worked

**Comment by space1worm** [comment]
@ferbragon   I have same issue ONLY when I use middleware, I have tried `serverComponentsExternalPackages: ['sequelize', 'sequelize-typescript'],`  but didn't fix it

**Reply by Esg450** [reply]
Also having this issue. It seems like it only gets triggered when attempting to use middleware.

**Reply by panda-geant** [reply]
Hi, I actually have the exact same problem. Did you find a solution ?

**Reply by Jan-T-Berg** [reply]
`const nextConfig = {
    experimental: {
        serverComponentsExternalPackages: ['sequelize', 'sequelize-typescript'],
    }
}

module.exports = nextConfig `


Theo helped

**Reply by panda-geant** [reply]
I tried it, but I never worked

**Reply by Jan-T-Berg** [reply]
How do you import the dependencies in the middelware? 

Can you write me a basic version with the error so that I can reproduce it? Preferably on a git repo

**Reply by panda-geant** [reply]
I import like this in middleware.ts

```
import { stackMiddlewares } from "@/src/middleware/stackMiddlewares";

import { authenticationMiddleware } from "@/src/middleware/authMiddleware";

const middlewares = [authenticationMiddleware];
export default stackMiddlewares(middlewares);
```

where stackMiddlewares is 

```
import { NextMiddleware, NextResponse } from "next/server";
import { MiddlewareFactory } from "@/src/middleware/types";

export function stackMiddlewares(functions: MiddlewareFactory[] = [], index = 0): NextMiddleware {
  const current = functions[index];
  if (current) {
    const next = stackMiddlewares(functions, index + 1);
    return current(next);
  }
  return () => NextResponse.next();
}
```

my auth middleware

```
import { NextFetchEvent, NextRequest } from "next/server";
import { MiddlewareFactory } from "./types";

export const authenticationMiddleware: MiddlewareFactory = (next) => {

  return async (request: NextRequest, _next: NextFetchEvent) => {
    console.log(request);

    // We will implement logic later, now, i will just add a dummy user
    
    return next(request, _next);
  };

};
```

and types =

```
import { NextMiddleware } from "next/server";

export type MiddlewareFactory = (middleware: NextMiddleware) => NextMiddleware;
```


When I delete the file "middleware.ts",I don't have the error.

**Comment by alessandrojcm** [comment]
I am having the exact same error with Next 14.0.4, it happens only when running from a middleware. Server components and route handlers work well, has anyone managed to solve this?

My `serverComponentsExternalPackages` has `'sequelize', 'oracledb', 'sharp'`

UPDATE: Ok so the file that causes the error from Sequelize uses the `util` module from Nodejs. Since middlewares do not [support the Node.js runtime](https://github.com/vercel/next.js/discussions/46722) I am afraid we are dead in the water until that changes.

**Reply by ayushmina** [reply]
Have you solved this error?

**Reply by alessandrojcm** [reply]
No, I actually ran into it today again but with the `redis` package. It's just an artificial limitation imposed by the Next.js team and until they decide to either lift it or implement these Node APIs in edge we are dead in the water.

**Reply by oxedom** [reply]
So would the best approach be creating an extemeral rotute endpoint and querying it from the middleware?

**Comment by saitrend** [comment]
I had exactly the same issue with Next.js 14.x. It was on Docker which was also running MySQL container. The connection was processed through the sequelize ORM which was throwing exactly the same issue. 

The solution:
- call to the sequelize connection should be processed before calling any other sequelize functions.
- In the Next config file add sequelize in array serverComponentsExternalPackages 
- And, assign MySQL2 in the dialectModules.

See the video https://youtu.be/jO8y1vHsyD0?si=kmSvKqBDBdMDqYSJ for more details.

I hope this solves everyone's problem.

**Comment by cagan123-alt** [comment]
I found the solution so to help i wanted to explain it. I hope it helps 

within the edge-runtime you can't use nodejs runtime required modules like jsonwebtoken or sequelize 

which means u cant use this packages without any update coming to nextjs that gives middlewares nodejs runtime support or you can implement it in nextjs repo. 

But what you can do in an edge-runtime is sending fetch requests.

so for example if you make a new endpoint like 
/api/secure/login/route.ts
or 
/api/authHandler/login/route.ts 

or whatever you prefer you can actually handle it there and do whatever you want in 1 fetch request 

example code 
```
import NextAuth, { NextAuthConfig,User } from "next-auth"
import Credentials from "next-auth/providers/credentials"
import { ofetch } from "ofetch"
export const BASE_PATH = "/api/auth"
const authOptions:NextAuthConfig = {
  providers: [
    Credentials({
      credentials: {
        email: {
          label:"Email",type:"text",
        },
        password: {
          label:"Password"
        },
      },
      async authorize(credentials):Promise<User | null> {        
        const user = await ofetch(`${process.env.MODE === "dev" ? process.env.NEXT_PUBLIC_DEV_URL : process.env.NEXT_PUBLIC_PROD_URL}/api/secure/login`,{
          method:"POST",
          headers:{
            "Content-Type":"application/json"
          },
          body:JSON.stringify(credentials)
        })
       

        //you can extend it to your needs 
        return user ? 
        {id:user.id,email:user.email} : null
      }
    }),
  ],
  basePath: BASE_PATH,
  secret:process.env.NEXTAUTH_SECRET

}
export const { handlers, signIn, signOut, auth } = NextAuth(authOptions)
```


my endpoint which is /api/secure/login/route.ts

```
import getAdminUserWiaCredentials from "@/_lib/adminUser/getAdminUserWiaCredentials";
import { NextRequest } from "next/server";

export async function POST(req:NextRequest) {
    const { email, password } = await req.json()    
    
    const user = await getAdminUserWiaCredentials(email, password);
    if (user) {
       return Response.json(user);
    }
    return Response.error();
}

```

and getAdminUserWiaCredentials is

```
import parseDataValues from "@/_utils/parsers/parseDataValues";
import validatePassword from "@/_utils/secure/validatePassword";
import { AdminUser } from "@/models";
import { AdminUserAttributes, AdminUserAttributesWithoutPassword } from "@/models/adminUsers/adminUser.types";

export default async function getAdminUserWiaCredentials(email: string, pwHash: string): Promise<AdminUserAttributesWithoutPassword | null> {
    try{
        const withPassword = parseDataValues(await AdminUser.findOne({
            where:{
                email:email,
            }
        })) as AdminUserAttributes
        if(!withPassword) return null        
         const isMatch = await validatePassword(pwHash, withPassword.password)         
         if(!isMatch) return null
        const {password,...withoutPassword} = withPassword
        return withoutPassword
    }
    catch(err){
        console.log(err)
        return null
    }
}
```

You can extend it as you prefer.

Until node runtime is supported within the middleware .This is the best way i found to solve the issue.

If you have another way to do it i would approaciate the share.
Hope this helps

**Comment by leerob** [comment, marked as answer]
Good news everyone, **Node.js Middleware support landed in Next.js 15.2** (experimental). This is now ready for testing. For any issues you run into, please open new issues so we can track and address before marking it as stable. Thank you!

https://nextjs.org/blog/next-15-2#nodejs-middleware-experimental
